From 428fc09088a95d56840daa9e0274854bc7541927 Mon Sep 17 00:00:00 2001
From: Elad Nachman <enachman@marvell.com>
Date: Sun, 18 May 2025 14:02:58 +0000
Subject: [PATCH 1/2] drivers: i2c: fix after kdump crash

Spurious interrupts left while i2c controller still
at RX or TX state would try to access the RX or TX
buffer pointer, which is NULL.
Add check to verify buffer pointer is not NULL before
reading or writing the buffer for additional TX or RX
operations.

Signed-off-by: Elad Nachman <enachman@marvell.com>
---
 patch/0018-i2c-mv64xxx-kdump-sonic.patch | 24 ++++++++++++++++++++++++
 patch/series                             |  1 +
 2 files changed, 25 insertions(+)
 create mode 100644 patch/0018-i2c-mv64xxx-kdump-sonic.patch

diff --git a/patch/0018-i2c-mv64xxx-kdump-sonic.patch b/patch/0018-i2c-mv64xxx-kdump-sonic.patch
new file mode 100644
index 0000000..94711b1
--- /dev/null
+++ b/patch/0018-i2c-mv64xxx-kdump-sonic.patch
@@ -0,0 +1,24 @@
+--- a/drivers/i2c/busses/i2c-mv64xxx.c	2025-05-14 17:43:42.421398200 +0300
++++ b/drivers/i2c/busses/i2c-mv64xxx.c	2025-06-10 22:07:10.137321500 +0300
+@@ -414,14 +415,18 @@
+ 		break;
+ 
+ 	case MV64XXX_I2C_ACTION_SEND_DATA:
+-		writel(drv_data->msg->buf[drv_data->byte_posn++],
+-			drv_data->reg_base + drv_data->reg_offsets.data);
++		if (drv_data->msg && drv_data->msg->buf)
++			writel(drv_data->msg->buf[drv_data->byte_posn++],
++				drv_data->reg_base + drv_data->reg_offsets.data);
+ 		writel(drv_data->cntl_bits,
+ 			drv_data->reg_base + drv_data->reg_offsets.control);
+ 		break;
+ 
+ 	case MV64XXX_I2C_ACTION_RCV_DATA:
+-		drv_data->msg->buf[drv_data->byte_posn++] =
++		if (drv_data->msg && drv_data->msg->buf)
++			drv_data->msg->buf[drv_data->byte_posn++] =
++				readl(drv_data->reg_base + drv_data->reg_offsets.data);
++		else
+ 			readl(drv_data->reg_base + drv_data->reg_offsets.data);
+ 		writel(drv_data->cntl_bits,
+ 			drv_data->reg_base + drv_data->reg_offsets.control);
diff --git a/patch/series b/patch/series
index 407b21d..c857b28 100755
--- a/patch/series
+++ b/patch/series
@@ -183,6 +183,7 @@ cisco-npu-disable-other-bars.patch
 0015-arm64-dts-marvell-Add-Supermicro-SSE-G3748-board.patch
 0016-arm64-dts-marvell-Add-Wistron-ES-1227-54TS-board.patch
 0017-Extend-driver-to-support-XMC-XM25QH256C-device.patch
+0018-i2c-mv64xxx-kdump-sonic.patch
 
 # amd-pensando elba support
 0000-Add-support-for-the-TI-TPS53659.patch
-- 
2.25.1

